src/agreement/lease-process-pool.test.ts

```diff
@@ -0,0 +1,344 @@
+      when(proposalPool.acquire()).thenResolve({} as OfferProposal);
+      when(proposalPool.remove(_)).thenResolve();
+      when(marketModule.createLease(_, _)).thenReturn({} as LeaseProcess);
+
+      const fakeAgreement = { getDto: () => ({}) } as Agreement;
+      when(agreementApi.proposeAgreement(_))
+        .thenCall(() => new Promise((resolve) => setTimeout(() => resolve(fakeAgreement), 50)))
+        .thenCall(
+          () => new Promise((_, reject) => setTimeout(() => reject(new Error("Failed to propose agreement")), 50)),
+        );
+      // all further calls will return the same error
+
+      const pool = getLeasePool({ min: 3 });
+
+      await expect(pool.ready(AbortSignal.timeout(60))).rejects.toThrow(
+        "Could not create enough lease processes to reach the minimum pool size in time",
+      );
+      expect(pool.getAvailable()).toBe(1);
+      // first loop 3 times, then 2 times
+      verify(proposalPool.acquire()).times(5);
+    });
+  });
+  describe("acquire()", () => {
+    it("takes a random lease process from the pool if none have activities", async () => {
+      when(agreementApi.getAgreement(_)).thenResolve({ getState: () => "Approved" } as Agreement);
+      const pool = getLeasePool({ min: 3 });
+      const lease1 = getMockLeaseProcess();
+      const lease2 = getMockLeaseProcess();
+      const lease3 = getMockLeaseProcess();
+      pool["lowPriority"].add(lease1);
+      pool["lowPriority"].add(lease2);
+      pool["lowPriority"].add(lease3);
+
+      const leaseProcess = await pool.acquire();
+      expect(pool.getBorrowed()).toBe(1);
+      expect(pool.getAvailable()).toBe(2);
+      expect([lease1, lease2, lease3]).toContain(leaseProcess);
+    });
+    it("prioritizes lease processes from high priority pool", async () => {
+      when(agreementApi.getAgreement(_)).thenResolve({ getState: () => "Approved" } as Agreement);
+      const pool = getLeasePool({ min: 3 });
+      const lease1 = getMockLeaseProcess();
+      const lease2 = getMockLeaseProcess();
+      const lease3 = getMockLeaseProcess();
+      pool["lowPriority"].add(lease1);
+      pool["highPriority"].add(lease2);
+      pool["lowPriority"].add(lease3);
+
+      const leaseProcess = await pool.acquire();
+      expect(pool.getBorrowed()).toBe(1);
+      expect(pool.getAvailable()).toBe(2);
+      expect(leaseProcess).toBe(lease2);
+    });
+    it("creates a new lease process if none are available", async () => {
+      LeaseProcessPool.prototype["createNewLeaseProcess"] = jest
+        .fn()
+        .mockResolvedValue({ agreement: { id: "1", getDto: () => ({}) } } as LeaseProcess);
+
+      const pool = getLeasePool({ min: 3 });
+      expect(pool.getSize()).toBe(0);
+      await pool.acquire();
+      expect(pool.getSize()).toBe(1);
+      expect(pool.getBorrowed()).toBe(1);
+      expect(pool.getAvailable()).toBe(0);
+    });
+    it("waits for a lease to become available when the pool is full", async () => {
```

Comment:
What was really neat about the generic-pool was that this waiting time could be also controlled with a timeout. Will we need `acquireTimeoutSec` in golem-js? Considering the use case from TaskExecutor: I have 100 tasks to run and only 5 workers (lease processes). I think it would be beneficial to be able to specify that the acquire timeout to short-circuit the waiting time. Users could specify `acquire({ timeoutSec: 15 })` to fail the operation if they didn't get the resource needed.

---

This reminds me about another feature that the generic-pool had: the eviction mechanism - if a resource was sitting idle for a configured amount of time, it was destroyed. It was a nice auto-scaling feature of generic-pool.

---

I wouldn't like to bloat this PR with additional requirements. I'm going to put them in our backlog to implement next.